File: 2022/swift/05.swift
import Foundation

let filename = "./input/05.in"

func main() {
    do {
        let contents = try String(contentsOfFile: filename).components(separatedBy: "\n\n").map {
            group in
            group.split(separator: "\n").map { String($0) }
        }

        let size = contents[0].count

        let generateStack = {
            contents[0][0 ... size - 2].map {
                row in
                Array(row).enumerated().filter {
                    $0.0 % 4 == 1
                }.map { $0.1 }
            }.enumerated().reduce([[Character]](repeating: [Character](repeating: " ", count: size), count: size)) {
                acc, curr in
                var next = acc
                let (y, value) = curr

                value.enumerated().forEach {
                    if $0.1 != " " {
                        next[$0.0][size - y - 2] = $0.1
                    }
                }

                return next
            }.map {
                $0.filter { $0 != " " }
            }
        }

        let instructions = contents[1].map {
            move in

            let spec = move.split(separator: " ").map { Int($0) }.filter { $0 != nil }.map { $0! }

            let qty = spec[0]
            let from = spec[1] - 1
            let to = spec[2] - 1

            return (qty, from, to)
        }

        let crane9000 = instructions.reduce(generateStack()) {
            stacks, move in

            var update = stacks

            let (qty, from, to) = move

            let start = stacks[from].count - qty

            stacks[from][start...].reversed().forEach {
                update[to].append($0)
            }

            update[from].removeLast(qty)

            return update
        }.map { $0[$0.count - 1] }

        print("Part one:", String(crane9000))

        let crane9001 = instructions.reduce(generateStack()) {
            stacks, move in

            var update = stacks

            let (qty, from, to) = move

            let start = stacks[from].count - qty

            stacks[from][start...].forEach {
                update[to].append($0)
            }

            update[from].removeLast(qty)

            return update
        }.map { $0[$0.count - 1] }

        print("Part two:", String(crane9001))

    } catch {
        print(error)
    }
}

main()

File: 2022/deno/05.ts
const input = await Deno.readTextFile("./input/05.in");

const [initial, instructions] = input
  .split("\n\n")
  .map((block) => block.split("\n"));

const reg = /(\s{4}|[A-Z{1}])/g;

const getStacks = () =>
  initial
    .slice(0, -1) // labels at bottom
    .map((row) => row.match(reg) || [])
    .reduce<string[][]>(
      (cols, row, _) => {
        row.forEach((cell, index) => {
          if (cell.trim()) {
            cols[index] = [cell, ...cols[index]];
          }
        });
        return cols;
      },
      Array.from({ length: 9 }, () => [] as string[])
    );

const unwind =
  (transport: (source: string[], start: number) => string[]) =>
  (stacks: string[][], move: string) => {
    const [qty, from, to] = move
      .split(" ")
      .map(Number)
      .filter((x) => !isNaN(x));

    const selected = stacks[from - 1];
    const start = selected.length - qty;

    stacks[to - 1].push(...transport(selected, start));

    selected.length = selected.length - qty;

    return stacks;
  };

/**
 * Part One
 */

const crane9000 = unwind((selected: string[], start: number) => {
  return selected.slice(start).reverse();
});

console.log(
  "Part one:",
  instructions
    .reduce(crane9000, getStacks())
    .map((stack) => stack[stack.length - 1])
    .join("")
);

/**
 * Part Two
 */

const crane9001 = unwind((selected: string[], start: number) => {
  return selected.slice(start);
});

console.log(
  "Part two:",
  instructions
    .reduce(crane9001, getStacks())
    .map((stack) => stack[stack.length - 1])
    .join("")
);
